src/paho/mqtt/client.py

```diff
@@ -490,9 +487,14 @@ def __init__(self, client_id="", clean_session=True, userdata=None,
         if not clean_session and (client_id == "" or client_id is None):
             raise ValueError('A client id must be provided if clean session is False.')
 
-        self._transport = transport
+        if transport.lower() not in ('websockets', 'tcp'):
+            raise ValueError('transport must be "websockets" or "tcp", not %s' % transport)
+        self._transport = transport.lower()
         self._protocol = protocol
-        self._userdata = userdata
+        if isinstance(userdata, collections.Iterable) and not isinstance(userdata, basestring):
+            self._userdata = collections.deque(userdata)
```

Comment:
docstrings indicate that `userdata` can be any type, but then it's treated inside `_do_publish` as an iterable with a `pop` method.  What should happen here?  Should `userdata` be enforced to be a deque and raise a `TypeError` if it's not an iterable that can be coerced into one?